### DISCUSSION TITLE
How to enable semantic highlighting with gopls?

### DISCUSSION THREAD

**Discussion by kcaashish** [LazyVim/LazyVim#471] [Q&A] [answered]
I am using neovim nightly. I have enabled `sematicTokens` for golang through `gopls`, but I don't see any difference. 

```lua
{
    "neovim/nvim-lspconfig",
    opts = {
      setup = {
        gopls = function(_, opts)
          local ih = require("inlay-hints")
          require("lazyvim.util").on_attach(function(client, bufnr)
            if client.name == "gopls" then
              ih.on_attach(client, bufnr)
            end
          end)
          opts.settings = {
            gopls = {
              semanticTokens = true,
              analyses = {
                unusedparams = true,
              },
              staticcheck = true,
              hints = {
                assignVariableTypes = true,
                compositeLiteralFields = true,
                compositeLiteralTypes = true,
                constantValues = true,
                functionTypeParameters = true,
                parameterNames = true,
                rangeVariableTypes = true,
              },
            },
          }
        end,
      },
    },
  }
```

How it is highlighted in VS Code:
![image](https://user-images.githubusercontent.com/31175326/226570936-447d6d06-f7ca-4e7f-80de-f85fd03c06ee.png)

How it appears in neovim right now:
![image](https://user-images.githubusercontent.com/31175326/226571504-75a55027-8dd0-4134-ab58-4300251152bb.png)

Notice `http` for the difference in how they appear.

**Comment by folke** [comment]
semantic tokens are enabled by default on nightly. Obviously it's impossible to have the exact same hl in Neovim and vscode

**Reply by kcaashish** [reply]
In this case, `ctx`, `request` and `rules` are highlighted `orange` shouldn't subsequent mentions of the variables inside the function have same highlighting? I am trying to understand if there is a different setting or something.
![image](https://user-images.githubusercontent.com/31175326/226575885-7f57bcad-d7a3-4135-9e89-7c7f2c16a91f.png)

**Reply by folke** [reply]
It does for lua and typescript. Not sure if golang supports that.

You can use `:Inspect` when on the var name to see if any semantic tokens were added by the lsp

**Reply by kcaashish** [reply]
Thank you! I just checked through `:Inspect` and semantic tokens aren't in fact added by gopls as they are in Lua. I will search further for Go.

**Reply by amaanq** [reply]
gopls is pretty broken w/ semantic highlighting at the moment, I had the same issue.

**Reply by kcaashish** [reply]
I found a work around for now. https://github.com/golang/go/issues/54531#issuecomment-1464982242

**Reply by amaanq** [reply]
wow great, someone should submit a PR to fix this ;) 

(PS: mark your comment as the solution/solved)

**Comment by kcaashish** [comment, marked as answer]
If someone ends up here, follow this to get semantic token hl with gopls: https://github.com/golang/go/issues/54531#issuecomment-1464982242

So my config changed to:

```lua
{
    "neovim/nvim-lspconfig",
    opts = {
      setup = {
        gopls = function(_, opts)
          local ih = require("inlay-hints")
          require("lazyvim.util").on_attach(function(client, bufnr)
            if client.name == "gopls" then
              ih.on_attach(client, bufnr)
              -- workaround for gopls not supporting semanticTokensProvider
              -- https://github.com/golang/go/issues/54531#issuecomment-1464982242
              if not client.server_capabilities.semanticTokensProvider then
                local semantic = client.config.capabilities.textDocument.semanticTokens
                client.server_capabilities.semanticTokensProvider = {
                  full = true,
                  legend = {
                    tokenTypes = semantic.tokenTypes,
                    tokenModifiers = semantic.tokenModifiers,
                  },
                  range = true,
                }
              end
            end
          end)
          opts.settings = {
            gopls = {
              semanticTokens = true,
              analyses = {
                unusedparams = true,
              },
              staticcheck = true,
              hints = {
                assignVariableTypes = true,
                compositeLiteralFields = true,
                compositeLiteralTypes = true,
                constantValues = true,
                functionTypeParameters = true,
                parameterNames = true,
                rangeVariableTypes = true,
              },
            },
          }
        end,
      },
    },
  }
```
